01-messaging.md

```diff
@@ -385,6 +398,466 @@ every message maximally).
+        "value": 0,
+        "bytes": "fd",
+        "exp_error": "unexpected EOF"
+    },
+    {
+        "name": "four byte no read",
+        "value": 0,
+        "bytes": "fe",
+        "exp_error": "unexpected EOF"
+    },
+    {
+        "name": "eight byte no read",
+        "value": 0,
+        "bytes": "ff",
+        "exp_error": "unexpected EOF"
+    }
+]
+```
+
+### BigSize Encoding Tests
+
+The following is an example of how to execute the BigSize encoding tests.
+```golang
+func testWriteVarInt(t *testing.T, test varIntTest) {
+        var (
+                w   bytes.Buffer
+                buf [8]byte
+        )
+        err := tlv.WriteVarInt(&w, test.Value, &buf)
+        if err != nil {
+                t.Fatalf("unable to encode %d as varint: %v",
+                        test.Value, err)
+        }
+
+        if bytes.Compare(w.Bytes(), test.Bytes) != 0 {
+                t.Fatalf("expected bytes: %v, got %v",
+                        test.Bytes, w.Bytes())
+        }
+}
+```
+
+A correct implementation should pass against the following test vectors:
+```json
+[
+    {
+        "name": "zero",
+        "value": 0,
+        "bytes": "00"
+    },
+    {
+        "name": "one byte high",
+        "value": 252,
+        "bytes": "fc"
+    },
+    {
+        "name": "two byte low",
+        "value": 253,
+        "bytes": "fd00fd"
+    },
+    {
+        "name": "two byte high",
+        "value": 65535,
+        "bytes": "fdffff"
+    },
+    {
+        "name": "four byte low",
+        "value": 65536,
+        "bytes": "fe00010000"
+    },
+    {
+        "name": "four byte high",
+        "value": 4294967295,
+        "bytes": "feffffffff"
+    },
+    {
+        "name": "eight byte low",
+        "value": 4294967296,
+        "bytes": "ff0000000100000000"
+    },
+    {
+        "name": "eight byte high",
+        "value": 18446744073709551615,
+        "bytes": "ffffffffffffffffff"
+    }
+]
+```
+
+## Appendix B: Type-Length-Value Test Vectors
+
+The following tests assume that two separate TLV namespaces exist: n1 and n2.
+
+The n1 namespace supports the following TLV types:
+
+1. tlvs: `n1`
+2. types:
+   1. type: 1 (`tlv1`)
+   2. data:
+     * [`tu64`:`amount_msat`]
+   1. type: 2 (`tlv2`)
+   2. data:
+     * [`short_channel_id`:`scid`]
+   1. type: 3 (`tlv3`)
+   2. data:
+     * [`point`:`node_id`]
+     * [`u64`:`amount_msat_1`]
+     * [`u64`:`amount_msat_2`]
+   1. type: 254 (`tlv4`)
+   2. data:
+     * [`u16`:`cltv_delta`]
+
+The n2 namespace supports the following TLV types:
+
+1. tlvs: `n2`
+2. types:
+   1. type: 0 (`tlv1`)
+   2. data:
+     * [`tu64`:`amount_msat`]
+   1. type: 11 (`tlv2`)
+   2. data:
+     * [`tu32`:`cltv_expiry`]
+
+### TLV Decoding Failures
+
+The following TLV streams in any namespace should trigger a decoding failure:
+
+1. Invalid stream: 0xfd
+2. Reason: type truncated
+
+1. Invalid stream: 0xfd01
+2. Reason: type truncated
+
+1. Invalid stream: 0xfd0001 00
```

Comment:
not a blocker, but i get this error without the trailing zero. do we want go for the most minimal test that exercises the error?

Reply:
I think the trailing `00` is useful. Without it you have two reasons for failing: not minimally-encoded type and missing length.

Reply:
That’s fair, tbh I haven’t done the math on whether the order of operations matters too much here

Reply:
That’s fair, tbh I haven’t done the math on whether the order of operations matters too much here

Reply:
sgtm. i think there are a few cases where OOO matters, but don't think that's the case with this testcase